test_pthread_proxying is flaky

This failure comes through every now and then:

```
======================================================================
FAIL: test_pthread_proxying (test_core.core0.test_pthread_proxying)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.12/unittest/case.py", line 634, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
       ^^^^^^^^
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 928, in resulting_test
    return func(self, *args)
           ^^^^^^^^^^^^^^^^^
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 385, in decorated
    f(self, *args, **kwargs)
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 700, in metafunc
    f(self, *args, **kwargs)
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/test_core.py", line 2606, in test_pthread_proxying
    self.do_run_in_out_file_test('pthread/test_pthread_proxying.c', interleaved_output=False)
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 1980, in do_run_in_out_file_test
    output = self._build_and_run(srcfile, expected, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 2029, in _build_and_run
    self.assertContained(o, js_output, regex=regex)
  File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 1687, in assertContained
    self.fail("Expected to find '%s' in '%s', diff:\n\n%s\n%s" % (
  File "/usr/lib/python3.12/unittest/case.py", line 715, in fail
    raise self.failureException(msg)
AssertionError: Expected to find 'Testing async proxying
running widget 1 on main
running widget 2 on looper
running widget 3 on returner
Testing sync proxying
running widget 4 on looper
running widget 5 on returner
Testing sync_with_ctx proxying
running widget 6 on looper
running widget 7 on returner
Testing callback proxying
running widget 8 on main
running widget 9 on looper
running widget 10 on returner
Testing callback_with_ctx proxying
running widget 11 on main
running widget 12 on looper
running widget 13 on returner
Testing promise proxying
running widget 14 on worker
running widget 15 on looper
running widget 16 on returner
Testing promise_with_ctx proxying
running widget 17 on worker
running widget 18 on looper
running widget 19 on returner
Testing tasks queue growth
Testing proxying queue growth
work
work
work
work
done
' in 'Testing async proxying
running widget 1 on main
running widget 2 on looper
running widget 3 on returner
Testing sync proxying
running widget 4 on looper
running widget 5 on returner
Testing sync_with_ctx proxying
running widget 6 on looper
running widget 7 on returner
Testing callback proxying
running widget 8 on main
running widget 9 on looper
running widget 10 on returner
Testing callback_with_ctx proxying
running widget 11 on main
running widget 12 on looper
running widget 13 on returner
Testing promise proxying
running widget 14 on worker
running widget 15 on looper
running widget 16 on returner
Testing promise_with_ctx proxying
running widget 17 on unknown
running widget 18 on looper
running widget 19 on returner
Testing tasks queue growth
Testing proxying queue growth
work
work
work
work
done
looper_main
returner_main
quitting looper
returner_main
returner_main
program exited (with status: 0), but keepRuntimeAlive() is set (counter=0) due to an async operation, so halting execution but not exiting the runtime or preventing further async execution (you can use emscripten_force_exit, if you want to force a true shutdown)
', diff:

--- expected
+++ actual
@@ -21,7 +21,7 @@
 running widget 15 on looper
 running widget 16 on returner
 Testing promise_with_ctx proxying
-running widget 17 on worker
+running widget 17 on unknown
 running widget 18 on looper
 running widget 19 on returner
 Testing tasks queue growth
@@ -31,4 +31,10 @@
 work
 work
 done
+looper_main
+returner_main
+quitting looper
+returner_main
+returner_main
+program exited (with status: 0), but keepRuntimeAlive() is set (counter=0) due to an async operation, so halting execution but not exiting the runtime or preventing further async execution (you can use emscripten_force_exit, if you want to force a true shutdown)
```
which suggests some kind of a race condition in proxying. (a lot of other pthread proxying tests are already marked flaky)

> This failure comes through every now and then:
> 
> ```
> ======================================================================
> FAIL: test_pthread_proxying (test_core.core0.test_pthread_proxying)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/usr/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
>     yield
>   File "/usr/lib/python3.12/unittest/case.py", line 634, in run
>     self._callTestMethod(testMethod)
>   File "/usr/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
>     if method() is not None:
>        ^^^^^^^^
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 928, in resulting_test
>     return func(self, *args)
>            ^^^^^^^^^^^^^^^^^
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 385, in decorated
>     f(self, *args, **kwargs)
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 700, in metafunc
>     f(self, *args, **kwargs)
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/test_core.py", line 2606, in test_pthread_proxying
>     self.do_run_in_out_file_test('pthread/test_pthread_proxying.c', interleaved_output=False)
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 1980, in do_run_in_out_file_test
>     output = self._build_and_run(srcfile, expected, **kwargs)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 2029, in _build_and_run
>     self.assertContained(o, js_output, regex=regex)
>   File "/home/clb/buildbot/h12dsi-linux-mint22/emscripten_linux_x64/build/emscripten/main/test/common.py", line 1687, in assertContained
>     self.fail("Expected to find '%s' in '%s', diff:\n\n%s\n%s" % (
>   File "/usr/lib/python3.12/unittest/case.py", line 715, in fail
>     raise self.failureException(msg)
> AssertionError: Expected to find 'Testing async proxying
> running widget 1 on main
> running widget 2 on looper
> running widget 3 on returner
> Testing sync proxying
> running widget 4 on looper
> running widget 5 on returner
> Testing sync_with_ctx proxying
> running widget 6 on looper
> running widget 7 on returner
> Testing callback proxying
> running widget 8 on main
> running widget 9 on looper
> running widget 10 on returner
> Testing callback_with_ctx proxying
> running widget 11 on main
> running widget 12 on looper
> running widget 13 on returner
> Testing promise proxying
> running widget 14 on worker
> running widget 15 on looper
> running widget 16 on returner
> Testing promise_with_ctx proxying
> running widget 17 on worker
> running widget 18 on looper
> running widget 19 on returner
> Testing tasks queue growth
> Testing proxying queue growth
> work
> work
> work
> work
> done
> ' in 'Testing async proxying
> running widget 1 on main
> running widget 2 on looper
> running widget 3 on returner
> Testing sync proxying
> running widget 4 on looper
> running widget 5 on returner
> Testing sync_with_ctx proxying
> running widget 6 on looper
> running widget 7 on returner
> Testing callback proxying
> running widget 8 on main
> running widget 9 on looper
> running widget 10 on returner
> Testing callback_with_ctx proxying
> running widget 11 on main
> running widget 12 on looper
> running widget 13 on returner
> Testing promise proxying
> running widget 14 on worker
> running widget 15 on looper
> running widget 16 on returner
> Testing promise_with_ctx proxying
> running widget 17 on unknown
> running widget 18 on looper
> running widget 19 on returner
> Testing tasks queue growth
> Testing proxying queue growth
> work
> work
> work
> work
> done
> looper_main
> returner_main
> quitting looper
> returner_main
> returner_main
> program exited (with status: 0), but keepRuntimeAlive() is set (counter=0) due to an async operation, so halting execution but not exiting the runtime or preventing further async execution (you can use emscripten_force_exit, if you want to force a true shutdown)
> ', diff:
> 
> --- expected
> +++ actual
> @@ -21,7 +21,7 @@
>  running widget 15 on looper
>  running widget 16 on returner
>  Testing promise_with_ctx proxying
> -running widget 17 on worker
> +running widget 17 on unknown
>  running widget 18 on looper
>  running widget 19 on returner
>  Testing tasks queue growth
> @@ -31,4 +31,10 @@
>  work
>  work
>  done
> +looper_main
> +returner_main
> +quitting looper
> +returner_main
> +returner_main
> +program exited (with status: 0), but keepRuntimeAlive() is set (counter=0) due to an async operation, so halting execution but not exiting the runtime or preventing further async execution (you can use emscripten_force_exit, if you want to force a true shutdown)
> ```
> 
> which suggests some kind of a race condition in proxying. (a lot of other pthread proxying tests are already marked flaky)

--- a/system/lib/pthread/pthread_create.c
+++ b/system/lib/pthread/pthread_create.c
@@ -171,6 +171,11 @@ int __pthread_create(pthread_t* restrict res,
   struct pthread *new = (struct pthread*)offset;
   offset += sizeof(struct pthread);
 
+  // Assign the pthread_t object over immediately, so that by the time pthread_create_js()
+  // is dispatched to a pthread and the pthread main runs, the value will be visible to
+  // the thread to examine.
+  __atomic_store_n(res, new, __ATOMIC_SEQ_CST);
+
   new->map_base = block;
   new->map_size = size;
 
@@ -250,6 +255,10 @@ int __pthread_create(pthread_t* restrict res,
 
   int rtn = __pthread_create_js(new, &attr, entry, arg);
   if (rtn != 0) {
+    // Reset the pthread_t return value to zero (we assigned to it above,
+    // so by clearing it here we won't litter bits to caller)
+    __atomic_store_n(res, 0, __ATOMIC_SEQ_CST);
+
     if (!--libc.threads_minus_1) libc.need_locks = 0;
 
     // undo previous addition to the thread list
@@ -269,7 +278,6 @@ int __pthread_create(pthread_t* restrict res,
       self->prev,
       new);
 
-  *res = new;
   return 0;
 }
 

--- a/system/lib/pthread/pthread_create.c
+++ b/system/lib/pthread/pthread_create.c
@@ -171,11 +171,6 @@ int __pthread_create(pthread_t* restrict res,
   struct pthread *new = (struct pthread*)offset;
   offset += sizeof(struct pthread);
 
-  // Assign the pthread_t object over immediately, so that by the time pthread_create_js()
-  // is dispatched to a pthread and the pthread main runs, the value will be visible to
-  // the thread to examine.
-  __atomic_store_n(res, new, __ATOMIC_SEQ_CST);
-
   new->map_base = block;
   new->map_size = size;
 
@@ -252,6 +247,12 @@ int __pthread_create(pthread_t* restrict res,
   // by the time it returns the thread could be running and we
   // want libc.need_locks to be set from the moment it starts.
   if (!libc.threads_minus_1++) libc.need_locks = 1;
+
+  // Assign the pthread_t object over immediately, so that by the time pthread_create_js()
+  // is dispatched to a pthread and the pthread main runs, the value will be visible to
+  // the thread to examine.
+  // Use __atomic_store_n() instead of a_store() to avoid splicing the pointer.
+  __atomic_store_n(res, new, __ATOMIC_SEQ_CST);
 
   int rtn = __pthread_create_js(new, &attr, entry, arg);
   if (rtn != 0) {

--- a/test/code_size/test_codesize_minimal_pthreads.json
+++ b/test/code_size/test_codesize_minimal_pthreads.json
@@ -1,10 +1,10 @@
 {
   "a.out.js": 7659,
   "a.out.js.gz": 3776,
-  "a.out.nodebug.wasm": 19592,
-  "a.out.nodebug.wasm.gz": 9060,
-  "total": 27251,
-  "total_gz": 12836,
+  "a.out.nodebug.wasm": 19599,
+  "a.out.nodebug.wasm.gz": 9063,
+  "total": 27258,
+  "total_gz": 12839,
   "sent": [
     "a (memory)",
     "b (emscripten_get_now)",

--- a/test/code_size/test_codesize_minimal_pthreads_memgrowth.json
+++ b/test/code_size/test_codesize_minimal_pthreads_memgrowth.json
@@ -1,10 +1,10 @@
 {
   "a.out.js": 8086,
   "a.out.js.gz": 3977,
-  "a.out.nodebug.wasm": 19593,
-  "a.out.nodebug.wasm.gz": 9061,
-  "total": 27679,
-  "total_gz": 13038,
+  "a.out.nodebug.wasm": 19600,
+  "a.out.nodebug.wasm.gz": 9064,
+  "total": 27686,
+  "total_gz": 13041,
   "sent": [
     "a (memory)",
     "b (emscripten_get_now)",

--- a/test/pthread/test_pthread_proxying_reduced_stress_test_case.c
+++ b/test/pthread/test_pthread_proxying_reduced_stress_test_case.c
@@ -1,0 +1,28 @@
+#include <emscripten/proxying.h>
+#include <emscripten.h>
+#include <stdio.h>
+
+pthread_t main_thread, looper, returner, worker, worker2;
+em_proxying_queue* proxy_queue = NULL;
+
+void* dummy(void* arg) { return NULL; }
+void dummy2(void* arg) {}
+
+void* worker2_main(void* arg) {
+  pthread_t self = pthread_self();
+  const char* name = pthread_equal(self, worker2) ? "worker2" : "unknown";
+  printf("pthread_self() == %s (%p)\n", name, (void*)self);
+  return NULL;
+}
+
+int main(int argc, char* argv[]) {
+  main_thread = pthread_self();
+  proxy_queue = em_proxying_queue_create();
+  pthread_create(&looper, NULL, dummy, NULL);
+  pthread_create(&returner, NULL, dummy, NULL);
+  emscripten_proxy_callback(proxy_queue, pthread_self(), dummy2, dummy2, NULL, NULL);
+  pthread_create(&worker, NULL, dummy, 0);
+  pthread_join(worker, NULL);
+  pthread_create(&worker2, NULL, worker2_main, 0);
+  pthread_join(worker2, NULL);
+}

--- a/test/test_core.py
+++ b/test/test_core.py
@@ -2581,7 +2581,6 @@ The current type of b is: 9
 
   @node_pthreads
   @also_with_modularize
-  @flaky('https://github.com/emscripten-core/emscripten/issues/25026')
   def test_pthread_proxying(self):
     if '-sMODULARIZE' in self.cflags:
       if self.get_setting('WASM') == 0:

--- a/test/test_stress.py
+++ b/test/test_stress.py
@@ -102,15 +102,13 @@ class stress(RunnerCore):
   @also_with_modularize
   @is_slow_test
   def test_stress_pthread_proxying(self):
-    self.skipTest('https://github.com/emscripten-core/emscripten/issues/25026')
     if '-sMODULARIZE' in self.cflags:
       if self.get_setting('WASM') == 0:
         self.skipTest('MODULARIZE + WASM=0 + pthreads does not work (#16794)')
       self.set_setting('EXPORT_NAME=ModuleFactory')
-    self.maybe_closure()
     self.set_setting('PROXY_TO_PTHREAD')
     if not self.has_changed_setting('INITIAL_MEMORY'):
       self.set_setting('INITIAL_MEMORY=32mb')
 
-    js_file = self.build('pthread/test_pthread_proxying.c')
-    self.parallel_stress_test_js_file(js_file, not_expected='running widget 17 on unknown', expected='running widget 17 on worker', assert_returncode=0)
+    js_file = self.build('pthread/test_pthread_proxying_reduced_stress_test_case.c')
+    self.parallel_stress_test_js_file(js_file, not_expected='pthread_self() == unknown', expected='pthread_self() == worker2', assert_returncode=0)

